lib/spack/spack/cmd/mirror.py

```diff
@@ -162,57 +172,76 @@ def mirror_create(args):
-
-        # Default name for directory is spack-mirror-<DATESTAMP>
-        directory = args.directory
-        if not directory:
-            timestamp = datetime.now().strftime("%Y-%m-%d")
-            directory = 'spack-mirror-' + timestamp
-
-        # Make sure nothing is in the way.
-        existed = os.path.isdir(directory)
-
-        # Actually do the work to create the mirror
-        present, mirrored, error = spack.mirror.create(
-            directory, specs, num_versions=args.versions_per_spec)
-        p, m, e = len(present), len(mirrored), len(error)
-
-        verb = "updated" if existed else "created"
-        tty.msg(
-            "Successfully %s mirror in %s" % (verb, directory),
-            "Archive stats:",
-            "  %-4d already present"  % p,
-            "  %-4d added"            % m,
-            "  %-4d failed to fetch." % e)
-        if error:
-            tty.error("Failed downloads:")
-            colify(s.cformat("{name}{@version}") for s in error)
-            sys.exit(1)
+                mirror_specs = spack.mirror.get_all_versions(specs)
+                mirror_specs.sort(
+                    key=lambda s: s.format("{name}{@version}").lower())
+        elif args.all_versions:
+            base_specs = list(specs)
+            specs = list()
+            for spec in base_specs:
+                specs.extend(Spec(x) for x in
+                             spec.package.possible_dependencies())
+            mirror_specs = spack.mirror.get_all_versions(specs)
+        else:
+            # If the user asked for dependencies, traverse spec DAG get them.
+            if args.dependencies:
+                new_specs = set()
+                for spec in specs:
+                    spec.concretize()
+                    for s in spec.traverse():
+                        new_specs.add(s)
+                specs = list(new_specs)
+
+            # Skip external specs, as they are already installed
+            external_specs = [s for s in specs if s.external]
+            specs = [s for s in specs if not s.external]
+
+            for spec in external_specs:
+                msg = 'Skipping {0} as it is an external spec.'
+                tty.msg(msg.format(spec.cshort_spec))
+
+            # Get concrete specs for each matching version of these specs.
+            mirror_specs = spack.mirror.get_matching_versions(
+                specs, num_versions=(args.versions_per_spec or 1))
+            for s in mirror_specs:
+                s.concretize()
+
+    # Default name for directory is spack-mirror-<DATESTAMP>
+    directory = args.directory
+    if not directory:
+        timestamp = datetime.now().strftime("%Y-%m-%d")
+        directory = 'spack-mirror-' + timestamp
+
+    # Make sure nothing is in the way.
+    existed = os.path.isdir(directory)
+
+    # Actually do the work to create the mirror
+    present, mirrored, error = spack.mirror.create(directory, mirror_specs)
```

Comment:
should create return a `MirrorStats` object?  I think you can even make this unpacking work if you add `__iter__` to `MirrorStats`

Reply:
I added a function to `MirrorStats` to fit the expected return type. I could see it being useful to return the object itself if we wanted to report more-detailed stats in the future but for now this seems sufficient.